File: lee/src/main/java/org/example/lee/题目/字典树/b_实现Trie前缀树.java
package org.example.lee.题目.字典树;

public class b_实现Trie前缀树 {

  class Trie {

    public Node root;

    public Trie() {
      root = new Node();
    }

    public void insert(String word) {
      int len = word.length(), i = 0;
      Node node = root;
      while (i < len) {
        char c = word.charAt(i);
        if (node.next[c - 'a'] == null) {
          node.next[c - 'a'] = new Node();
        }
        node = node.next[c - 'a'];
        i++;
      }
      node.canBeEnd = true;
    }

    public boolean search(String word) {
      Node tool = tool(word);
      if (tool != null) {
        return tool.canBeEnd;
      }
      return false;
    }

    public boolean startsWith(String prefix) {
      Node tool = tool(prefix);
      return tool != null;
    }

    private Node tool(String str) {
      int len = str.length(), i = 0;
      Node node = root;
      while (i < len) {
        char c = str.charAt(i);
        if (node.next[c - 'a'] == null) {
          return null;
        }
        node = node.next[c - 'a'];
        i++;
      }
      return node;
    }

    class Node {

      public Node[] next;
      public boolean canBeEnd;

      public Node() {
        next = new Node[26];
      }
    }
  }

}

File: lee/src/main/java/org/example/lee/题目/字典树/b_实现Trie前缀树.go
package 字典树

type Trie struct {
	root *Node
}

func Constructor() Trie {
	return Trie{root: NewNode()}
}

func (this *Trie) Insert(word string) {
	node := this.root
	for _, v := range word {
		if node.nodes[v-'a'] == nil {
			node.nodes[v-'a'] = NewNode()
		}
		node = node.nodes[v-'a']
	}
	node.canEnd = true
}

func (this *Trie) Search(word string) bool {
	tool, node := this.tool(word)
	return tool == true && node.canEnd == true
}

func (this *Trie) StartsWith(prefix string) bool {
	tool, _ := this.tool(prefix)
	return tool
}
func (this *Trie) tool(prefix string) (bool, *Node) {
	node := this.root
	for _, v := range prefix {
		if node.nodes[v-'a'] == nil {
			return false, node
		}
		node = node.nodes[v-'a']
	}
	return true, node
}
func NewNode() *Node {
	return &Node{
		nodes: make([]*Node, 26),
	}
}

type Node struct {
	nodes  []*Node
	canEnd bool
}
